apps/designer/app/dashboard/projects/create-rename-duplicate-delete.tsx

```diff
@@ -0,0 +1,276 @@
+        gap="1"
+      >
+        {description && (
+          <DialogDescription asChild>
+            <Paragraph>{description}</Paragraph>
+          </DialogDescription>
+        )}
+        <Label>{label}</Label>
+        <TextField
+          placeholder={placeholder}
+          name="title"
+          defaultValue={title}
+          state={errors ? "invalid" : undefined}
+          onChange={(event) => {
+            onChange?.({ title: event.currentTarget.value });
+          }}
+        />
+        <Box css={{ minHeight: theme.spacing["10"] }}>
+          {errors && <Text color="error">{errors}</Text>}
+        </Box>
+      </Flex>
+      <DialogActions>
+        {primaryButton}
+        <DialogClose asChild>
+          <Button color="ghost">Cancel</Button>
+        </DialogClose>
+      </DialogActions>
+    </form>
+  );
+};
+
+const trpc = createTrpcRemixProxy<DashboardProjectRouter>(dashboardProjectPath);
+
+const useCreateProject = () => {
+  const navigate = useNavigate();
+  const { submit, data } = trpc.create.useMutation();
+
+  useEffect(() => {
+    if (data === undefined || "errors" in data) {
+      return;
+    }
+    navigate(designerPath({ projectId: data.id }));
+  }, [data, navigate]);
+
+  return {
+    handleSubmit: submit,
+    errors: data && "errors" in data ? data.errors : undefined,
+  };
+};
+
+export const CreateProject = () => {
+  const { handleSubmit, errors } = useCreateProject();
+  return (
+    <Dialog
+      title="New Project"
+      trigger={<Button prefix={<PlusIcon />}>New Project</Button>}
+    >
+      <Content
+        onSubmit={handleSubmit}
+        errors={errors}
+        placeholder="New Project"
+        label="Project Title"
+        primaryButton={<Button type="submit">Create Project</Button>}
+      />
+    </Dialog>
+  );
+};
+
+const useRenameProject = ({
+  projectId,
+  onComplete,
+}: {
+  projectId: DashboardProject["id"];
+  onComplete: () => void;
+}) => {
+  const { submit, data } = trpc.rename.useMutation();
+
+  useEffect(() => {
+    if (data === undefined || "errors" in data) {
+      return;
+    }
+    onComplete();
+  }, [data, onComplete]);
+
+  const handleSubmit = ({ title }: { title: string }) => {
+    submit({ projectId, title });
+  };
+
+  return {
+    handleSubmit,
+    errors: data && "errors" in data ? data.errors : undefined,
+  };
+};
+
+export const RenameProject = ({
+  isOpen,
+  title,
+  projectId,
+  onComplete,
+  onOpenChange,
+}: {
+  isOpen: boolean;
+  title: string;
+  projectId: DashboardProject["id"];
+  onComplete: () => void;
+  onOpenChange: (isOpen: boolean) => void;
+}) => {
+  const { handleSubmit, errors } = useRenameProject({ projectId, onComplete });
+  return (
+    <Dialog title="Rename" isOpen={isOpen} onOpenChange={onOpenChange}>
+      <Content
+        onSubmit={handleSubmit}
+        errors={errors}
+        title={title}
+        label="Project Title"
+        primaryButton={<Button type="submit">Rename Project</Button>}
```

Comment:
Probably should make the `Button` pending while request is being handled. To have some indication that clicking the button had an effect, and to prevent user from sending another request.

Same in other forms.

Reply:
agre